### PR TITLE
top-level, lib: Remove platform attribute of platforms

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

First commit: take II of #107214 which for some reason became a mass rebuild.

Second commit: Simplify top level.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-burns** [comment]
I think this broke the tarball build on darwin which is blocking unstable

**Comment by Ericson2314** [comment]
I'll investigate.

File: nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix [hunk 1]
```diff
@@ -103,6 +103,6 @@ in
 
     system.build.installBootLoader = builder;
     system.boot.loader.id = "raspberrypi";
-    system.boot.loader.kernelFile = platform.kernelTarget;
```
Comment (ajs124) [inline nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:106]: `linux-kernel` does not exist in this scope. `platform` comes from `inherit (pkgs.stdenv.hostPlatform) platform;` and is now unused.

This does not evaluate.

Comment (Ericson2314) [reply nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:106]: Sorry, I will fix. How should I test this?

Comment (ajs124) [reply nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:106]: Good questions. `nix-build -E '(import ./. {}).nixos { boot.loader.raspberryPi.enable = true; boot.loader.grub.enable = false; fileSystems."/".device = "e"; }'` should work.

Comment (Ericson2314) [reply nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix:106]: See https://github.com/NixOS/nixpkgs/pull/110803

**Comment by samueldr** [comment]
Were those values intended to be entirely internal to Nixpkgs?

Should a compatibility shim be put in place to prevent external breakage?

Right now, I'm in a situation where I cannot just push the fix to Mobile NixOS because there are other breakage with cross-compilation that needs to be addressed. Applying the patch means I need an out-of-tree workaround to work with a known-good Nixpkgs. Not applying it means Hydra can't show the progress with regards to regressions.

As a broader topic, not really fit for here, what's the expected stability of Nixpkgs APIs? Should things just disappear like that? Should there always be a transition period where both are possible?

Sorry, I don't want to only rant here. I know the change is entirely made with good intentions in mind, and with Nixpkgs in mind.

I want to get the ball rolling about thinking how to make all the platform things less problematic to use external to Nixpkgs. I think there were other changes with 20.09->21.05, that'll bite a few individuals who stick to stable I'm sure.

**Pull request merged**